clippy_lints/src/methods/mod.rs

```diff
@@ -2049,6 +2050,31 @@ declare_clippy_lint! {
     "unnecessary calls to `to_owned`-like functions"
 }
 
+declare_clippy_lint! {
+    /// ### What it does
+    /// Checks for use of `.collect::<Vec<String>>().join("")` on iterators.
+    ///
+    /// ### Why is this bad?
+    /// `.collect::<String>()` is more concise and usually more performant
+    ///
+    /// ### Example
+    /// ```rust
+    /// let vector = vec!["hello",  "world"];
+    /// let output = vector.iter().map(|item| item.to_uppercase()).collect::<Vec<String>>().join("");
+    /// println!("{}", output);
+    /// ```
+    /// The correct use would be:
+    /// ```rust
+    /// let vector = vec!["hello",  "world"];
+    /// let output = vector.iter().map(|item| item.to_uppercase()).collect::<String>();
+    /// println!("{}", output);
+    /// ```
+    #[clippy::version = "pre 1.61.0"]
+    pub UNNECESSARY_JOIN,
+    perf,
```

Comment:
With the perf analysis you did in the Rust PR, I don't think we can put this lint in "perf". Loop unrolling usually gives a pretty big perf benefit. 

I'm happy to include this lint in Clippy, because it might improve code size and is easier to read. 

Maybe the different code generation needs to be fixed in rustc? But until then I would put it in `pedantic`. I don't want to put it in `style` and make it warn-by-default, because it could have a negative perf impact.

Please also add a `Known Problems` section to the lint documentation explaining that using `.collect::<String>` might prevent loop unrolling and thus have a negative perf impact.

Reply:
That sounds reasonable, will add this

Reply:
Done

Reply:
Not sure about the "most cases" part since logically it should be and is in the cases I've measured, but it's hard to know what usually happens. On the other hand, using wording like "can be" would be underselling it a bit IMO

Reply:
Your formulation sounds ideal!